### ISSUE TITLE
Using a Solidity library

### ISSUE THREAD

**Issue by hexonaut** [makerdao/dss-exec-lib#41]
I'm curious as to the design decision to use inherited delegate calls to DssExecLib as a contract instead of using it as a library. It seems like having the solidity compiler do the heavily lifting for that would be not only less code to manage, but reduce the chance of error by function signature typos. Solidity libraries can be dynamically linked as well to reduce the amount of code getting deployed into the spell.

**Comment by brianmcmichael** [comment]
Solidity libraries aren't really useful for modifying state. They're commonly used for things like safemath and other pure functions that take parameters and return results. The original plan was to use native libraries but we ended up not being able to modify system params though it.

**Comment by hexonaut** [comment]
Didn't realize that. How did it prevent you from doing that? Compiler error? Runtime revert?

**Comment by brianmcmichael** [comment]
No revert, which was really promising until we started running tests. It basically manipulates the state of the calling contract instead of the contract you're expecting to call. Now that I'm thinking of it, this might be very dangerous if called within the context of the pause. We probably need to add some checks in the spell tests to ensure that pause values are not changed from pre- to post- test unless explicitly called.

**Comment by hexonaut** [comment]
Would be interesting to look at the bytecode for the library. Are external calls in libraries using delegate calls or something instead of regular calls? Maybe there is a way to explicitly set it to use a call. I'll run some tests to see what's going on.

**Comment by hexonaut** [comment]
I wrote up a proof of concept to test using libraries: https://github.com/hexonaut/lib-test

If you run `./test.sh` on Kovan it updates the LibTest contract from the library just fine. I tested both the embedded use case as well as the dynamically linked one. I think it could be worthwhile to make this switch as it will remove the need to replicate the function calls in `DssAction`. Instead of calling functions like `libCall("setIlkDebtCeiling(address,bytes32,uint256)", vat(), ilk, amount);` we could just do `ExecLib.setIlkDebtCeiling(vat(), ilk, amount)` to get compiler safety.
